### PR TITLE
Vectorize Transform Vector node

### PR DESCRIPTION
**Pull request by harisreedhar**
Vectorized Matrix input of `Transform Vector` node.

Resolves: #1671

### CODE REVIEW

File: animation_nodes/nodes/vector/c_utils.pyx [hunk 1]
```diff
@@ -114,3 +114,12 @@ def calculateVectorDotProducts(Py_ssize_t amount,
     for i in range(amount):
         dotProducts.data[i] = dotVec3(vectors1.get(i), vectors2.get(i))
     return dotProducts
+
+def transformVirtualVectorList(Py_ssize_t amount,
+                        VirtualVector3DList vectors,
+                        VirtualMatrix4x4List matrices):
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/vector/c_utils.pyx:120]: Align those with the first argument.

File: animation_nodes/nodes/vector/transform_vector.py [hunk 2]
```diff
@@ -6,19 +6,27 @@ class TransformVectorNode(bpy.types.Node, AnimationNode):
...
     def create(self):
         self.newInput(VectorizedSocket("Vector", "useVectorList",
             ("Vector", "vector"),
             ("Vectors", "vectors", dict(dataIsModified = True))))
+        self.newInput(VectorizedSocket("Matrix", "useMatrixList",
+            ("Matrix", "matrix"), ("Matrices", "matrices")))
 
-        self.newInput("Matrix", "Matrix", "matrix")
-
-        self.newOutput(VectorizedSocket("Vector", "useVectorList",
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/vector/transform_vector.py:17]: VectorizedSocket can take multiple properties as input. `["useVectorList", "useMatrixList"]`. This should avoid the if statement.

File: animation_nodes/nodes/vector/transform_vector.py [hunk 3]
```diff
@@ -6,19 +6,27 @@ class TransformVectorNode(bpy.types.Node, AnimationNode):
     bl_label = "Transform Vector"
 
     useVectorList: VectorizedSocket.newProperty()
+    useMatrixList: VectorizedSocket.newProperty()
 
     def create(self):
         self.newInput(VectorizedSocket("Vector", "useVectorList",
             ("Vector", "vector"),
             ("Vectors", "vectors", dict(dataIsModified = True))))
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/vector/transform_vector.py:14]: `dict(dataIsModified = True)` is no longer needed.

File: animation_nodes/nodes/vector/transform_vector.py [hunk 4]
```diff
@@ -6,19 +6,27 @@ class TransformVectorNode(bpy.types.Node, AnimationNode):
...
 
     def getExecutionCode(self, required):
-        if self.useVectorList:
-            return "vectors.transform(matrix)"
+        if any((self.useVectorList, self.useMatrixList)):
+            if not self.useVectorList: yield "vectors = Vector3DList.fromValue(vector)"
+            if not self.useMatrixList: yield "matrices = Matrix4x4List.fromValue(matrix)"
+            yield "_vectors = VirtualVector3DList.create(vectors, (0,0,0))"
+            yield "_matrices = VirtualMatrix4x4List.create(matrices, Matrix.Identity(4))"
+            yield "amount = max(len(vectors), len(matrices))"
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/vector/transform_vector.py:29]: Use `getMaxRealLength`.

**Review by OmarEmaraDev — CHANGES_REQUESTED** [review]

File: animation_nodes/nodes/vector/transform_vector.py (hunk 2 above)
Comment (harisreedhar) [reply animation_nodes/nodes/vector/transform_vector.py:17]: By using this method there is an issue with socket update.

https://user-images.githubusercontent.com/46858047/116970318-5a6c7180-acd5-11eb-96ec-9a123ef63421.mp4

File: animation_nodes/nodes/vector/transform_vector.py (hunk 3 above)
Comment (harisreedhar) [reply animation_nodes/nodes/vector/transform_vector.py]: Done

File: animation_nodes/nodes/vector/transform_vector.py (hunk 4 above)
Comment (harisreedhar) [reply animation_nodes/nodes/vector/transform_vector.py]: Done

File: animation_nodes/nodes/vector/transform_vector.py (hunk 2 above)
Comment (OmarEmaraDev) [reply animation_nodes/nodes/vector/transform_vector.py:17]: Seems like a standard behavior for the vectorization selector. Perhaps this behavior should change globally then.

**Review by OmarEmaraDev — APPROVED** [review]
Thanks!

**Comment by Gerstmann-Bradley** [comment]
Sorry I haven't checked yet, but in description it only mentions "Transform Vector" which raises my concern.
I was asking the whole series: "Transform Matrix", "Transform Spline" both of high priority, and then Transform Vector, etc.
Are the other nodes being vectorized as well? If so it will be so nice. Thank you both.

**Comment by OmarEmaraDev** [comment]
Transform Matrix and Transforms Spline are already vectorized, no?

**Comment by Gerstmann-Bradley** [comment]
> Transform Matrix and Transforms Spline are already vectorized, no?

No, at least none of the series member is vectorized before this commit, and I still haven't checked this commit yet.
image

**Comment by OmarEmaraDev** [comment]
Okay right. One input is vectorized and the other isn't.

**Pull request merged**